### PR TITLE
Bypass Apply Requirements with --f

### PR DESCRIPTION
**Pull request by pcalley**
Adds the -f or --force option to atlantis apply comment commands:
```  -f, --force                 Force Atlantis to ignore apply requirments.```

Dynamic config related PR: https://github.com/lyft/atlantis-dynamic-config/pull/8

This will be disabled when first deployed.

### CODE REVIEW

**Comment by pcalley** [comment]
ptal #orchestration @nishkrishnan @msarvar @Aayyush

File: server/events/project_command_context_builder.go [hunk 1]
```diff
@@ -37,7 +37,7 @@ type ProjectCommandContextBuilder interface {
 		prjCfg valid.MergedProjectCfg,
 		commentFlags []string,
 		repoDir string,
-		automerge, deleteSourceBranchOnMerge, parallelApply, parallelPlan, verbose bool,
+		automerge, deleteSourceBranchOnMerge, parallelApply, parallelPlan, verbose, force bool,
```
Comment (msarvar) [inline server/events/project_command_context_builder.go:40]: I think these flags should be consolidated into a struct.

File: server/events/models/models.go [hunk 2]
```diff
@@ -411,6 +411,8 @@ type ProjectCommandContext struct {
 	User User
 	// Verbose is true when the user would like verbose output.
 	Verbose bool
+	// Force is true when the apply should ignore apply_requirements.
+	Force bool
```
Comment (msarvar) [inline server/events/models/models.go:415]: Let's name this ForceApply?

**Comment by msarvar** [comment]
great job, looks good overall. Couple nits

**Review by nishkrishnan — COMMENTED** [review]
Couple things. 

1.  Would be great if we could have a server flag that enables this since then we'd be able to move this upstream as well. Let's just do it as part of this PR since it should be pretty quick.
2. In the future, having the feature allocator be used in the least invasive way possible ensures you don't have to modify existing tests/ makes it easy to remove. Most obvious way to do this is using the decorator pattern.

```go
type SomeInterface interface {
  SomeFunc(...)
}


type SomeImpl struct {
}

func (s SomImpl) SomeFunc(...) {}

type AlternateImpl struct {}

func (s AlternateImpl) SomeFunc(...) {}

type FeatureAwareImpl struct{}

func (s FeatureAwareImpl) SomeFunc(...) {

   // call feature allocator and either route to AlternateImpl or SomeImpl

}
```

This makes it so that unit tests are scoped to the implementation and we can just remove the feature aware implementation once we're done without having to deal with fixing tests or editing implementation code.


In your case, one way to have done that would have been by only populating the `ProjectCommandContext` with the real value of force if the feature has been allocated and it could have been done as a delegate interface basically.

**Comment by msarvar** [comment]
yeah, we should do what @nishkrishnan suggested and gate the force apply's value behind the feature flag decorator.

**Comment by pcalley** [comment]
@nishkrishnan I'm definitely into the suggestion I'm not 100% I understand how to go about it. We have the https://github.com/lyft/atlantis/blob/da24e3c2646fdb5c38c2ce565d97ad7feb00b886/server/events/project_command_context_builder.go#L195 where we could move the FeatureAllocator logic and just pass the `force` value if the feature is enabled. I think that makes sense to me b/c it will simplify the project_command_runner.doApply() and then we wouldn't need to mock the feature allocator for the unit tests there. Is that sorta what you're thinking?

**Comment by msarvar** [comment]
You would create a new decorator struct, similar to
https://github.com/lyft/atlantis/blob/da24e3c2646fdb5c38c2ce565d97ad7feb00b886/server/events/project_command_runner.go#L113.
It can have the feature allocator, you can then set the force apply flag to
false if the feature is not enabled.

File: server/events/models/models.go (hunk 2 above)
Comment (pcalley) [reply server/events/models/models.go]: Renamed.

File: server/controllers/events/events_controller_e2e_test.go [hunk 3]
```diff
@@ -671,6 +674,7 @@ func setupE2E(t *testing.T, repoDir string) (events_controllers.VCSEventsControl
 	e2eGithubGetter := mocks.NewMockGithubPullGetter()
 	e2eGitlabGetter := mocks.NewMockGitlabMergeRequestGetter()
 	projectCmdOutputHandler := handlermocks.NewMockProjectCommandOutputHandler()
+	allocator := fmocks.NewMockAllocator()
```
[the remark this answers is not in the document]
Comment (pcalley) [reply server/controllers/events/events_controller_e2e_test.go]: SG. This got removed anyway from the changes above.

File: server/events/project_command_context_builder.go (hunk 1 above)
Comment (pcalley) [reply server/events/project_command_context_builder.go]: done

File: server/events/project_command_runner.go [hunk 4]
```diff
@@ -395,6 +405,9 @@ func (p *DefaultProjectCommandRunner) doApply(ctx models.ProjectCommandContext)
 		return "", "", fmt.Errorf("%s\n%s", err, strings.Join(outputs, "\n"))
 	}
 
+	if shouldAllocate && ctx.Force {
+		outputs = append(outputs, "WARNING: this apply run with --force option bypassing the apply requirements. This should only be used in an emergency.")
+	}
```
[the remark this answers is not in the document]
Comment (pcalley) [reply server/events/project_command_runner.go]: Cool I updated it so its a separate comment before the apply. Heres a screenshot. 
Screen Shot 2021-09-17 at 4 09 21 PM

**Comment by pcalley** [comment]
> Couple things.
> 
> 1. Would be great if we could have a server flag that enables this since then we'd be able to move this upstream as well. Let's just do it as part of this PR since it should be pretty quick.
> 2. In the future, having the feature allocator be used in the least invasive way possible ensures you don't have to modify existing tests/ makes it easy to remove. Most obvious way to do this is using the decorator pattern.
> 
> ```go
> type SomeInterface interface {
>   SomeFunc(...)
> }
> 
> 
> type SomeImpl struct {
> }
> 
> func (s SomImpl) SomeFunc(...) {}
> 
> type AlternateImpl struct {}
> 
> func (s AlternateImpl) SomeFunc(...) {}
> 
> type FeatureAwareImpl struct{}
> 
> func (s FeatureAwareImpl) SomeFunc(...) {
> 
>    // call feature allocator and either route to AlternateImpl or SomeImpl
> 
> }
> ```
> 
> This makes it so that unit tests are scoped to the implementation and we can just remove the feature aware implementation once we're done without having to deal with fixing tests or editing implementation code.
> 
> In your case, one way to have done that would have been by only populating the `ProjectCommandContext` with the real value of force if the feature has been allocated and it could have been done as a delegate interface basically.

Added the decorator logic here and updated the unit tests.
https://github.com/lyft/atlantis/blob/c4bf3c076faa6c3a3696b96449d3dfaf686fb476/server/events/project_command_runner.go#L156-L179

File: server/events/project_command_runner.go [hunk 5]
```diff
@@ -152,6 +153,31 @@ func (p *ProjectOutputWrapper) updateProjectPRStatus(commandName models.CommandN
 	return result
 }
 
+type FeatureAwareProjectCommandRunner struct {
+	ProjectCommandRunner
+	FeatureAllocator feature.Allocator
+}
+
+func (f *FeatureAwareProjectCommandRunner) Apply(ctx models.ProjectCommandContext) models.ProjectResult {
```
Comment (msarvar) [inline server/events/project_command_runner.go:161]: Can you add some  tests for this decorator as well?

Comment (pcalley) [reply server/events/project_command_runner.go:161]: Yeah totally just added some.

File: server/events/command_runner.go [hunk 6]
```diff
@@ -224,6 +224,13 @@ func (c *DefaultCommandRunner) RunCommentCommand(baseRepo models.Repo, maybeHead
 		ctx.Log.Err("Error running pre-workflow hooks %s. Proceeding with %s command.", err, cmd.Name.String())
 	}
 
+	if cmd.ForceApply {
+		warningMessage := "⚠️ WARNING ⚠️\n\n You have bypassed all apply requirements for this PR 🚀 . This can have unpredictable consequences 🙏🏽 and should only be used in an emergency 🆘 .\n\n 𝐓𝐡𝐢𝐬 𝐚𝐜𝐭𝐢𝐨𝐧 𝐰𝐢𝐥𝐥 𝐛𝐞 𝐚𝐮𝐝𝐢𝐭𝐞𝐝.\n"
+		if commentErr := c.VCSClient.CreateComment(baseRepo, pullNum, warningMessage, ""); commentErr != nil {
+			log.Err("unable to comment: %s", commentErr)
+		}
+	}
+
```
Comment (nishkrishnan) [inline server/events/command_runner.go:233]: Shouldn't this be wrapped in the feature flag as well? The feature flag call should be idempotent and quick so you could theoretically employ the decorator pattern again here and wrap this.

Comment (pcalley) [reply server/events/command_runner.go:234]: Makes sense. I need to add some unit tests quick but how does this look:  https://github.com/lyft/atlantis/pull/107/commits/bb3607c6e55a6217966a1c7b9bbc367bbf85e363

Comment (pcalley) [reply server/events/command_runner.go]: done

**Review by nishkrishnan — DISMISSED** [review]

**Review by nishkrishnan — DISMISSED** [review]

**Review by nishkrishnan — APPROVED** [review]

**Pull request merged**